File: src/app/profile-validation.ts
export type ProfileValidationErrors = Partial<{
  username: string;
  displayName: string;
  bio: string;
  location: string;
  website: string;
  avatarUrl: string;
}>;

const USERNAME_RE = /^[a-z0-9._]{3,24}$/;

export function normalizeProfileUsername(value: string): string {
  return value.trim().replace(/^@/, "").toLowerCase();
}

export function isValidWebsite(value: string): boolean {
  const website = value.trim();
  if (!website) return true;
  try {
    const candidate = /^https?:\/\//i.test(website) ? website : `https://${website}`;
    const parsed = new URL(candidate);
    const host = parsed.hostname.toLowerCase();
    return (parsed.protocol === "http:" || parsed.protocol === "https:") && (host.includes(".") || host === "localhost");
  } catch {
    return false;
  }
}

export function isSafeAvatarUrl(value?: string): boolean {
  const avatarUrl = value?.trim() ?? "";
  if (!avatarUrl) return true;
  if (avatarUrl.startsWith("data:")) return true;
  try {
    const parsed = new URL(avatarUrl);
    return (parsed.protocol === "http:" || parsed.protocol === "https:") && !!parsed.hostname;
  } catch {
    return false;
  }
}

export function validateProfilePatch(patch: Partial<{
  username: string;
  displayName: string;
  bio: string;
  location: string;
  website: string;
  avatarUrl: string;
}>): ProfileValidationErrors {
  const errors: ProfileValidationErrors = {};

  if (patch.username !== undefined) {
    const username = normalizeProfileUsername(patch.username ?? "");
    if (!username) {
      errors.username = "Pick a username.";
    } else if (!USERNAME_RE.test(username)) {
      errors.username = "Usernames are 3–24 characters: letters, numbers, dots and underscores.";
    }
  }

  if (patch.displayName !== undefined) {
    const displayName = (patch.displayName ?? "").trim();
    if (displayName.length > 40) {
      errors.displayName = "Display names can be up to 40 characters.";
    }
  }

  if (patch.bio !== undefined) {
    const bio = patch.bio ?? "";
    if (bio.length > 160) {
      errors.bio = "Bios can be up to 160 characters.";
    }
  }

  if (patch.location !== undefined) {
    const location = (patch.location ?? "").trim();
    if (location.length > 80) {
      errors.location = "Locations can be up to 80 characters.";
    }
  }

  if (patch.website !== undefined && !isValidWebsite(patch.website ?? "")) {
    errors.website = "Enter a valid website URL, like https://example.com.";
  }

  if (patch.avatarUrl !== undefined && !isSafeAvatarUrl(patch.avatarUrl)) {
    errors.avatarUrl = "Profile photos can only be uploaded from your device or kept as-is.";
  }

  return errors;
}

export function getProfileValidationError(patch: Partial<{ username: string; displayName: string; bio: string; location: string; website: string; avatarUrl: string; }>) {
  return Object.values(validateProfilePatch(patch))[0];
}

File: backend/app_legacy/profile_validation.py
from __future__ import annotations

import re
from urllib.parse import urlparse

USERNAME_RE = re.compile(r"[a-z0-9._]{3,24}")
DISPLAY_NAME_MAX = 40
BIO_MAX = 160


def normalize_username(value: str) -> str:
    return (value or "").strip().removeprefix("@").lower()


def validate_username(value: str) -> str:
    username = normalize_username(value)
    if not USERNAME_RE.fullmatch(username):
        raise ValueError("Usernames are 3-24 characters: letters, numbers, dots and underscores.")
    return username


def normalize_display_name(value: str) -> str:
    return " ".join((value or "").strip().split())


def validate_display_name(value: str) -> str:
    display_name = normalize_display_name(value)
    if not display_name:
        raise ValueError("Display name cannot be empty.")
    if len(display_name) > DISPLAY_NAME_MAX:
        raise ValueError("Display names can be up to 40 characters.")
    return display_name


def normalize_bio(value: str | None) -> str:
    return (value or "").strip()


def validate_bio(value: str | None) -> str:
    bio = normalize_bio(value)
    if len(bio) > BIO_MAX:
        raise ValueError("Bios can be up to 160 characters.")
    return bio


def normalize_website(value: str | None) -> str:
    website = (value or "").strip()
    if not website:
        return ""
    return website if re.match(r"^https?://", website, flags=re.IGNORECASE) else f"https://{website}"


def validate_website(value: str | None) -> str:
    website = normalize_website(value)
    if not website:
        return ""
    parsed = urlparse(website)
    host = (parsed.hostname or "").lower()
    if parsed.scheme.lower() not in {"http", "https"} or not host:
        raise ValueError("Enter a valid website URL, like https://example.com.")
    if "." not in host and host != "localhost":
        raise ValueError("Enter a valid website URL, like https://example.com.")
    return website


def normalize_avatar_url(value: str | None) -> str:
    return (value or "").strip()


def validate_avatar_url(value: str | None) -> str:
    avatar_url = normalize_avatar_url(value)
    if not avatar_url:
        return ""
    if avatar_url.startswith("data:"):
        return avatar_url
    parsed = urlparse(avatar_url)
    if parsed.scheme.lower() in {"http", "https"} and parsed.hostname:
        return avatar_url
    raise ValueError("Profile photos can only be uploaded from your device or kept as-is.")
